src/org/jgroups/util/SslContextFactory.java

```diff
@@ -251,4 +258,41 @@ private static void loadKeyStore(KeyStore ks, String keyStoreFileName, char[] ke
          Util.close(is);
       }
    }
+
+   public static Provider findProvider(String providerName, String serviceType, String algorithm) {
+      Provider[] providers = discoverSecurityProviders(Thread.currentThread().getContextClassLoader());
+      for (Provider provider : providers) {
+         if (providerName == null || providerName.equals(provider.getName())) {
+            Provider.Service providerService = provider.getService(serviceType, algorithm);
+            if (providerService != null) {
+               return provider;
+            }
+         }
+      }
+      return null;
+   }
+
+   public static Provider[] discoverSecurityProviders(ClassLoader classLoader) {
+      return PER_CLASSLOADER_PROVIDERS.computeIfAbsent(classLoader, cl -> {
+               // We need to keep them sorted by insertion order, since we want system providers first
+               Map<Class<? extends Provider>, Provider> providers = new LinkedHashMap<>();
+               for (Provider provider : Security.getProviders()) {
+                  providers.put(provider.getClass(), provider);
+               }
+               Iterator<Provider> loader = ServiceLoader.load(Provider.class, classLoader).iterator();
+               for (;;) {
+                  try {
+                     if (!loader.hasNext()) {
+                        return providers.values().toArray(new Provider[0]);
+                     } else {
+                        Provider provider = loader.next();
+                        providers.putIfAbsent(provider.getClass(), provider);
+                     }
+                  } catch (ServiceConfigurationError ignored) {
+                     // explicitly ignored
+                  }
+               }
```

Comment:
```suggestion
               for (Provider provider : ServiceLoader.load(Provider.class, classLoader)) {
                  providers.putIfAbsent(provider.getClass(), provider);
               }
```

Reply:
We must call `.next()` and `.hasNext()` inside a try-catch, because they may throw an exception. The above is the only way it will work.